pkg/kwok/metrics/metrics.go

```diff
@@ -0,0 +1,340 @@
+					return fmt.Errorf("failed to evaluate labels: %w", err)
+				}
+				val := prometheus.NewCounterFunc(
+					prometheus.CounterOpts{
+						Name:        m.Name,
+						Help:        m.Help,
+						ConstLabels: labels,
+					},
+					func() float64 {
+						return float64(counter.Load())
+					},
+				)
+
+				err = h.registry.Register(val)
+				if err != nil {
+					return fmt.Errorf("failed to register counter %q: %w", m.Name, err)
+				}
+				return nil
+			}
+
+			eval, err := h.environment.Compile(m.Value)
+			if err != nil {
+				return fmt.Errorf("failed to compile metric %q: %w", metric.Name, err)
+			}
+
+			var once sync.Once
+			updateFunc := func() error {
+				nodeInfo, ok := h.getNodeInfo(metric.Name)
+				if !ok {
+					return fmt.Errorf("failed to find node %q", metric.Name)
+				}
+
+				var err error
+				once.Do(func() {
+					err = registerFunc(nodeInfo.Node)
+				})
+				if err != nil {
+					return err
+				}
+
+				result, err := eval.EvaluateMetric(nodeInfo.Node)
+				if err != nil {
+					return fmt.Errorf("failed to evaluate metric %q: %w", m.Name, err)
+				}
+				counter.Store(int64(result))
+				return nil
+			}
+			h.updates = append(h.updates, updateFunc)
+		case KindHistogram:
+			buckets := make([]float64, 0, len(m.Buckets))
+			for _, b := range m.Buckets {
+				f, err := strconv.ParseFloat(b.Le, 64)
+				if err != nil {
+					return fmt.Errorf("failed to parse bucket %q: %w", b.Le, err)
+				}
+				buckets = append(buckets, f)
+			}
+			var val prometheus.Histogram
+
+			registerFunc := func(node *corev1.Node) error {
+				labels, err := h.createLabels(preparedLabels, node)
+				if err != nil {
+					return fmt.Errorf("failed to evaluate labels: %w", err)
+				}
+				val = prometheus.NewHistogram(
+					prometheus.HistogramOpts{
+						Name:        m.Name,
+						Help:        m.Help,
+						Buckets:     buckets,
+						ConstLabels: labels,
+					},
+				)
+				err = h.registry.Register(val)
+				if err != nil {
+					return fmt.Errorf("failed to register histogram %q: %w", m.Name, err)
+				}
+
+				return nil
+			}
+
+			bucketEvals := make(map[float64]*cel.Evaluator)
+			var infEval *cel.Evaluator
+			for _, b := range m.Buckets {
+				eval, err := h.environment.Compile(b.Value)
+				if err != nil {
+					return fmt.Errorf("failed to compile metric %q: %w", metric.Name, err)
+				}
+				if b.Le == "+Inf" {
+					infEval = eval
+					continue
+				}
+				le, err := strconv.ParseFloat(b.Le, 64)
+				if err != nil {
+					return fmt.Errorf("failed to parse bucket's le value %q: %w", b.Le, err)
+				}
+				bucketEvals[le] = eval
+			}
+
+			var once sync.Once
+			updateFunc := func() error {
+				nodeInfo, ok := h.getNodeInfo(metric.Name)
+				if !ok {
+					return fmt.Errorf("failed to find node %q", metric.Name)
+				}
+
+				var err error
+				once.Do(func() {
+					err = registerFunc(nodeInfo.Node)
+				})
+				if err != nil {
+					return err
+				}
+
+				for le, eval := range bucketEvals {
```

Comment:
Is this the logic that should be applied? Once this is done, I think it would be good for the complete review
Thank you for improving this PR @wzshiming !